worker/src/backgroundMigrations/migrateObservationsFromPostgresToClickhouse.ts

```diff
@@ -0,0 +1,165 @@
+import { IBackgroundMigration } from "./IBackgroundMigration";
+import { clickhouseClient, logger } from "@langfuse/shared/src/server";
+import { parseArgs } from "node:util";
+import { prisma, Prisma } from "@langfuse/shared/src/db";
+import { env } from "../env";
+import { Observation } from "@prisma/client";
+
+async function addTemporaryColumnIfNotExists() {
+  const columnExists = await prisma.$queryRaw<{ column_exists: boolean }[]>(
+    Prisma.sql`
+      SELECT EXISTS (
+        SELECT 1
+        FROM information_schema.columns
+        WHERE table_name = 'observations'
+        AND column_name = 'tmp_migrated_to_clickhouse'
+      ) AS column_exists;
+    `,
+  );
+  if (!columnExists[0]?.column_exists) {
+    await prisma.$executeRaw`ALTER TABLE observations ADD COLUMN tmp_migrated_to_clickhouse BOOLEAN DEFAULT FALSE;`;
+    logger.info("Added temporary column tmp_migrated_to_clickhouse");
+  } else {
+    logger.info(
+      "Temporary column tmp_migrated_to_clickhouse already exists. Continuing...",
+    );
+  }
+}
+
+export default class MigrateObservationsFromPostgresToClickhouse
+  implements IBackgroundMigration
+{
+  private isAborted = false;
+
+  async validate(
+    args: Record<string, unknown>,
+  ): Promise<{ valid: boolean; invalidReason: string | undefined }> {
+    if (!env.CLICKHOUSE_URL) {
+      return {
+        valid: false,
+        invalidReason: "Clickhouse URL must be configured to perform migration",
+      };
+    }
+    return { valid: true, invalidReason: undefined };
+  }
+
+  async run(args: Record<string, unknown>): Promise<void> {
+    logger.info(
+      `Migrating observations from postgres to clickhouse with ${JSON.stringify(args)}`,
+    );
+
+    const maxRowsToProcess = Number(args.maxRowsToProcess ?? Infinity);
+    const batchSize = Number(args.batchSize ?? 5000);
+    const maxDate = new Date((args.maxDate as string) ?? new Date());
+
+    await addTemporaryColumnIfNotExists();
+
+    let processedRows = 0;
+    while (!this.isAborted && processedRows < maxRowsToProcess) {
+      const observations = await prisma.$queryRaw<
+        Array<Observation>
+      >(Prisma.sql`
+        SELECT id, trace_id, project_id, type, parent_observation_id, start_time, end_time, name, metadata, level, status_message, version, input, output, unit, completion_start_time, internal_model_id, prompt_id, input_cost, output_cost, total_cost, created_at, updated_at
+        FROM observations
+        WHERE tmp_migrated_to_clickhouse = FALSE AND created_at <= ${maxDate}
+        LIMIT ${batchSize};
+      `);
+      if (observations.length === 0) {
+        logger.info("No more observations to migrate. Exiting...");
+        break;
+      }
+
+      await clickhouseClient.insert({
+        table: "observations",
+        values: observations.map((observation) => ({
+          ...observation,
```

Comment:
Let's chat about this offline. On the observation side this may need some thought.
Me may also need to check whether our "data quality" script needs any updates after those updates have been applied.